### DISCUSSION TITLE
`UseMutationOptions` generic type variables

### DISCUSSION THREAD

**Discussion by mkosir** [TanStack/query#3581] [General] [answered]
Hey 👋 
I guess more of TypeScript related topic, but still, `UseMutationOptions` is defined as:
```
export interface UseMutationOptions<TData = unknown, TError = unknown, TVariables = void, TContext = unknown>
extends Omit<MutationObserverOptions<TData, TError, TVariables, TContext>, '_defaulted' | 'variables'> {
}
```
I created a wrapper that whenever mutation is created, onError callback is handled, doing as:
```
export type UseMutationOptionsExtended<
    TData = unknown,
    TError = unknown,
    TVariables = void,
    TContext = unknown
> = RequiredField<UseMutationOptions<TData, TError, TVariables, TContext>, 'onError'>;
```
Wondering if there is a cleaner way to extend it, where passing generics type variables `TData, TError, TVariables, TContext` seems a bit redundant? Not sure if it would be possible to create some (exported) common generic in a library itself.

**Comment by TkDodo** [comment, marked as answer]
I think you need to pass all generics.

**Reply by dimmasyusuf** [reply]
I think it would be better if I could use it like this:

```
UseQueryOptions<ApiResponse>
UseMutationOptions<ApiResponse> // or UseMutationOptions<ApiRequest>

```

That way, it’s simpler and more straightforward.
